From 11c4b90b5d6b6b3d801488249c0ea8d27fa46038 Mon Sep 17 00:00:00 2001
From: "Gabriel F. T. Gomes" <gagomes@suse.de>
Date: Mon, 8 Jun 2020 10:41:30 -0300
Subject: [PATCH] Expose the status of locks used in malloc and dlopen

When applying live-patches, libpulp uses calloc, dlopen, and dlsym from
the context of a signal handler.  However, since these functions use
locks, they are AS-Unsafe, and calling them from a signal handler might
lead to a deadlock.  To avoid the deadlock, libpulp takes a two-steps
approach:

  1. Stop all threads in the target process (hijacking);
  2. Test that all relevant locks are free.

Testing the locks must be a non-blocking operation, otherwise, the
deadlock would occur while trying to avoid it.  This patch adds
__libpulp_dlopen_checks and __libpulp_malloc_checks to ld.so and libc,
respectively, which are non-blocking functions to test for the locks
used by calloc, dlopen, and dlsym.
---
 elf/Versions                                  |  4 ++
 elf/dl-misc.c                                 | 41 +++++++++++++++++
 malloc/Versions                               |  4 ++
 malloc/arena.c                                | 44 +++++++++++++++++++
 nptl/nptl-init.c                              |  1 +
 sysdeps/generic/ldsodefs.h                    |  4 ++
 sysdeps/nptl/libc-lockP.h                     | 13 ++++--
 sysdeps/unix/sysv/linux/x86_64/64/ld.abilist  |  2 +
 .../unix/sysv/linux/x86_64/64/libc.abilist    |  2 +
 9 files changed, 112 insertions(+), 3 deletions(-)

diff --git a/elf/Versions b/elf/Versions
index e65f2fac20..c0f59185e8 100644
--- a/elf/Versions
+++ b/elf/Versions
@@ -74,4 +74,8 @@ ld {
     # Set value of a tunable.
     __tunable_get_val;
   }
+  GLIBC_LIBPULP1 {
+    # Libpulp checks to avoid deadlock in AS-Unsafe functions.
+    __libpulp_dlopen_checks;
+  }
 }
diff --git a/elf/dl-misc.c b/elf/dl-misc.c
index c469b5a365..ebb4d746f0 100644
--- a/elf/dl-misc.c
+++ b/elf/dl-misc.c
@@ -444,3 +444,44 @@ _dl_strtoul (const char *nptr, char **endptr)
 
   return result;
 }
+
+/* Non-blocking function to test the locks used by dlopen and friends.
+
+   When applying live-patches, libpulp uses dlopen and dlsym from the
+   context of a signal handler.  However, since dlopen and dlsym use
+   locks, they are AS-Unsafe, and calling them from a signal handler
+   might lead to a deadlock.  To avoid the deadlock, libpulp takes a
+   two-steps approach:
+
+     1. Stop all threads in the target process (hijacking);
+     2. Test that all relevant locks are free.
+
+   Testing the locks must be a non-blocking operation, otherwise, the
+   deadlock would occur while trying to avoid it.
+
+   The following function tests that the locks used by dlopen and dlsym
+   are free, without blocking.
+
+   Return 1 if any of the locks are blocked, and 0 if all free.  */
+int
+__libpulp_dlopen_checks (void)
+{
+  int read;
+  int write;
+
+  /* Try and get the locks used in dlopen and dlsym.  */
+  read = __rtld_lock_trylock_recursive (GL(dl_load_lock));
+  write = __rtld_lock_trylock_recursive (GL(dl_load_write_lock));
+
+  /* Always release them.  */
+  if (read == 0)
+    __rtld_lock_unlock_recursive (GL(dl_load_lock));
+  if (write == 0)
+    __rtld_lock_unlock_recursive (GL(dl_load_write_lock));
+
+  /* If unable to get any of the locks, return 1; otherwise, 0.  */
+  if (read | write)
+    return 1;
+  return 0;
+}
+rtld_hidden_def (__libpulp_dlopen_checks)
diff --git a/malloc/Versions b/malloc/Versions
index 2357cff3da..1989e821fc 100644
--- a/malloc/Versions
+++ b/malloc/Versions
@@ -93,4 +93,8 @@ libc {
     __libc_alloc_buffer_copy_string;
     __libc_alloc_buffer_create_failure;
   }
+  GLIBC_LIBPULP1 {
+    # Libpulp checks to avoid deadlock in AS-Unsafe functions.
+    __libpulp_malloc_checks;
+  }
 }
diff --git a/malloc/arena.c b/malloc/arena.c
index dc14fae152..f3c160dc54 100644
--- a/malloc/arena.c
+++ b/malloc/arena.c
@@ -996,6 +996,50 @@ arena_thread_freeres (void)
 }
 text_set_element (__libc_thread_subfreeres, arena_thread_freeres);
 
+/* Non-blocking function to test the locks used by malloc and friends.
+
+   When applying live-patches, libpulp uses calloc from the context of a
+   signal handler.  However, since calloc uses locks, it is AS-Unsafe,
+   and calling it from a signal handler might lead to a deadlock.  To
+   avoid the deadlock, libpulp takes a two-steps approach:
+
+     1. Stop all threads in the target process (hijacking);
+     2. Test that all relevant locks are free.
+
+   Testing the locks must be a non-blocking operation, otherwise, the
+   deadlock would occur while trying to avoid it.
+
+   The following function tests that the locks used by calloc are free,
+   without blocking.  More specifically, it iterates over all arenas,
+   checking their locking state, as well as it checks the state of
+   list_lock and free_list_lock.
+
+   Return 1 if any of the locks are blocked, and 0 if all free.  */
+int
+__libpulp_malloc_checks (void)
+{
+  mstate arena;
+
+  /* Return 1 if any of the lists of arenas is locked. */
+  if (free_list_lock)
+    return 1;
+  if (list_lock)
+    return 1;
+
+  /* The circular linked list of arenas can be acessed from the main
+     arena, which is also part of the list.  So, go over all nodes and
+     stop when back at the main arena.  Return 1, if any of the arenas
+     is locked.  */
+  arena = &main_arena;
+  do {
+    if (arena->mutex)
+      return 1;
+    arena = arena->next;
+  } while (arena != &main_arena);
+
+  return 0;
+}
+
 /*
  * Local variables:
  * c-basic-offset: 2
diff --git a/nptl/nptl-init.c b/nptl/nptl-init.c
index 29216077a2..d8313c66f2 100644
--- a/nptl/nptl-init.c
+++ b/nptl/nptl-init.c
@@ -443,6 +443,7 @@ __pthread_initialize_minimal_internal (void)
   /* Make __rtld_lock_{,un}lock_recursive use pthread_mutex_{,un}lock,
      keep the lock count from the ld.so implementation.  */
   GL(dl_rtld_lock_recursive) = (void *) __pthread_mutex_lock;
+  GL(dl_rtld_trylock_recursive) = (void *) __pthread_mutex_trylock;
   GL(dl_rtld_unlock_recursive) = (void *) __pthread_mutex_unlock;
   unsigned int rtld_lock_count = GL(dl_load_lock).mutex.__data.__count;
   GL(dl_load_lock).mutex.__data.__count = 0;
diff --git a/sysdeps/generic/ldsodefs.h b/sysdeps/generic/ldsodefs.h
index 4508365871..fb8f4f04c3 100644
--- a/sysdeps/generic/ldsodefs.h
+++ b/sysdeps/generic/ldsodefs.h
@@ -371,6 +371,7 @@ struct rtld_global
 #if defined SHARED && defined _LIBC_REENTRANT \
     && defined __rtld_lock_default_lock_recursive
   EXTERN void (*_dl_rtld_lock_recursive) (void *);
+  EXTERN int (*_dl_rtld_trylock_recursive) (void *);
   EXTERN void (*_dl_rtld_unlock_recursive) (void *);
 #endif
 
@@ -1090,6 +1091,9 @@ extern void _dl_non_dynamic_init (void) internal_function;
 /* Used by static binaries to check the auxiliary vector.  */
 extern void _dl_aux_init (ElfW(auxv_t) *av) internal_function;
 
+/* Used by libpulp to test for internal locks in signal handlers.  */
+extern int __libpulp_dlopen_checks (void);
+rtld_hidden_proto (__libpulp_dlopen_checks)
 
 __END_DECLS
 
diff --git a/sysdeps/nptl/libc-lockP.h b/sysdeps/nptl/libc-lockP.h
index a43e62db4d..af9f83ac77 100644
--- a/sysdeps/nptl/libc-lockP.h
+++ b/sysdeps/nptl/libc-lockP.h
@@ -202,9 +202,6 @@ typedef pthread_key_t __libc_key_t;
 #define __libc_rwlock_trywrlock(NAME) \
   __libc_maybe_call (__pthread_rwlock_trywrlock, (&(NAME)), 0)
 
-#define __rtld_lock_trylock_recursive(NAME) \
-  __libc_maybe_call (__pthread_mutex_trylock, (&(NAME).mutex), 0)
-
 /* Unlock the named lock variable.  */
 #if IS_IN (libc) || IS_IN (libpthread)
 # define __libc_lock_unlock(NAME) \
@@ -220,18 +217,28 @@ typedef pthread_key_t __libc_key_t;
 # define __rtld_lock_default_lock_recursive(lock) \
   ++((pthread_mutex_t *)(lock))->__data.__count;
 
+# define __rtld_lock_default_trylock_recursive(lock) \
+  ++((pthread_mutex_t *)(lock))->__data.__count; \
+  return 0;
+
 # define __rtld_lock_default_unlock_recursive(lock) \
   --((pthread_mutex_t *)(lock))->__data.__count;
 
 # define __rtld_lock_lock_recursive(NAME) \
   GL(dl_rtld_lock_recursive) (&(NAME).mutex)
 
+# define __rtld_lock_trylock_recursive(NAME) \
+  GL(dl_rtld_trylock_recursive) (&(NAME).mutex)
+
 # define __rtld_lock_unlock_recursive(NAME) \
   GL(dl_rtld_unlock_recursive) (&(NAME).mutex)
 #else
 # define __rtld_lock_lock_recursive(NAME) \
   __libc_maybe_call (__pthread_mutex_lock, (&(NAME).mutex), 0)
 
+# define __rtld_lock_trylock_recursive(NAME) \
+  __libc_maybe_call (__pthread_mutex_trylock, (&(NAME).mutex), 0)
+
 # define __rtld_lock_unlock_recursive(NAME) \
   __libc_maybe_call (__pthread_mutex_unlock, (&(NAME).mutex), 0)
 #endif
diff --git a/sysdeps/unix/sysv/linux/x86_64/64/ld.abilist b/sysdeps/unix/sysv/linux/x86_64/64/ld.abilist
index 07cab4bd7a..861962ed19 100644
--- a/sysdeps/unix/sysv/linux/x86_64/64/ld.abilist
+++ b/sysdeps/unix/sysv/linux/x86_64/64/ld.abilist
@@ -9,3 +9,5 @@ GLIBC_2.2.5 realloc F
 GLIBC_2.3 GLIBC_2.3 A
 GLIBC_2.3 __tls_get_addr F
 GLIBC_2.4 GLIBC_2.4 A
+GLIBC_LIBPULP1 GLIBC_LIBPULP1 A
+GLIBC_LIBPULP1 __libpulp_dlopen_checks F
diff --git a/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist b/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
index 8bff2b225b..58749d2550 100644
--- a/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
+++ b/sysdeps/unix/sysv/linux/x86_64/64/libc.abilist
@@ -2176,3 +2176,5 @@ GLIBC_2.9 dup3 F
 GLIBC_2.9 epoll_create1 F
 GLIBC_2.9 inotify_init1 F
 GLIBC_2.9 pipe2 F
+GLIBC_LIBPULP1 GLIBC_LIBPULP1 A
+GLIBC_LIBPULP1 __libpulp_malloc_checks F
-- 
2.28.0

